File: 86.partition-list.java
/*
 * @lc app=leetcode id=86 lang=java
 *
 * [86] Partition List
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        if (head == null || head.next == null) return head;
        ListNode hl = new ListNode(-1), pl = hl;
        ListNode hg = new ListNode(-1), pg = hg;
        while (head != null) {
            if (head.val < x) {
                pl.next = head;
                pl = head;
            } else {
                pg.next = head;
                pg = head;
            }
            head = head.next;
        }
        pl.next = hg.next;
        pg.next = null;
        return hl.next;
    }
}

File: go/86.partition-list.go
/*
 * @lc app=leetcode id=86 lang=golang
 *
 * [86] Partition List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
	l1, l2 := new(ListNode), new(ListNode)
	h1, h2 := l1, l2
	for head != nil {
		tmp := head.Next
		head.Next = nil
		if head.Val < x {
			l1.Next = head
			l1 = l1.Next
		} else {
			l2.Next = head
			l2 = l2.Next
		}
		head = tmp
	}
	l1.Next = h2.Next
	return h1.Next
}
// @lc code=end

